### ISSUE TITLE
cups 2.4.0: shared remote printers not appearing/working

### ISSUE THREAD

**Issue by tgurr** [OpenPrinting/cups#306]
Small local LAN, remote host running latest Fedora with cups-2.3.3op2-11.fc35 and having the printer shared via the Web-UI configuration option.

On the "client" having setup avahi/mdns everything is working fine with cups-2.3.3_p2, the remote printer automatically appears, plain stock default configuration files whipped with cups, no manual or any changes done with the webinterface made.

After the upgrade to cups 2.4.0, also plain stock default cups configuration, the printer doesn't appear anymore and I get errors in the logs `/var/log/cups/error_log` every few seconds:

```
[...]
E [08/Dec/2021:19:17:25 +0100] [Client 984] Returning HTTP Verboten for CUPS-Add-Modify-Printer (ipp://localhost/printers/EPSON_XP_2100_Series_remote) from localhost
E [08/Dec/2021:19:17:35 +0100] [Client 986] Returning HTTP Verboten for CUPS-Add-Modify-Printer (ipp://localhost/printers/EPSON_XP_2100_Series_remote) from localhost
E [08/Dec/2021:19:17:45 +0100] [Client 988] Returning HTTP Verboten for CUPS-Add-Modify-Printer (ipp://localhost/printers/EPSON_XP_2100_Series_remote) from localhost
E [08/Dec/2021:19:17:55 +0100] [Client 990] Returning HTTP Verboten for CUPS-Add-Modify-Printer (ipp://localhost/printers/EPSON_XP_2100_Series_remote) from localhost
[...]
```
`Verboten` probably translates to `Forbidden`.

Just downgrading back to 2.3.0_p2 makes things work again so I'm kind of clueless as I've never had to make any manual changes and things just worked for me in the past and a diff between the configs of `cupsd.conf` of both versions also don't give any clue as there were only minor changes and none in regards to any policies:

```
-Browsing On
+Browsing Yes

+# Timeout after cupsd exits if idle (applied only if cupsd runs on-demand - with -l)
+IdleExitTimeout 60
```

hence I suspect some kind of bug.

**Comment by zdohnal** [comment]
Hi @tgurr ,

thank you for reporting the issue!

Just to be clear - you have problems on the client or on the server? I'm not sure if I understood where you'd installed 2.4.0.

Can you see the queue from server on mDNS communication? You can check via ```avahi-browse -avrt``` on the client.

And to debug this further, it would be great if you could give me debug logs from the client when the forbidden message appears. From the snippet you've shared I'm not sure when the rejection happens, but since you don't see the printer, my guess is it happens during print queue listing, so ```lpstat -e``` should hit the issue and it shouldn't return anything. 

Here are the steps:

```
$ sudo sed -i 's,LogLevel warn,LogLevel debug2,' /etc/cups/cupsd.conf
$ sudo systemctl restart cups
$ tail -f /var/log/cups/error_log &> log.txt &
$ lpstat -e
```
after lpstat finishes, kill the tail process. Please check if log.txt contains the errors you mentioned - if it does, do attach it here as attachment. Otherwise (since you reported the errors occurs regularly) catch the log till the error happens.

Thank you in advance!

**Comment by zdohnal** [comment]
The previous comment expects that you are using temporary queues - meaning your client's cupsd picks up queue advertised by mDNS during print dialog and the queue is not permanently installed - which means no cups-browsed is involved.

FTR I've tried the print queue sharing and discovering in 2.4.0 on two Fedora Rawhide machines (server+client) and it works. Steps were:

Server:
$ systemctl start cups
$ cupsctl --remote-any
$ lpadmin -p queue -v device-uri -m driver-uri -E

Client:
$ systemctl start cups
$ lpstat -e

**Comment by tgurr** [comment]
> Just to be clear - you have problems on the client or on the server? I'm not sure if I understood where you'd installed 2.4.0.

The problem is on the client where I upgraded to 2.4.0 (Exherbo Linux - packaging diff between our 2.3.3 and 2.4.0 version: https://git.exherbo.org/arbor.git/commit/?id=897656b18ff105d452e04b19143bbfb0e307e4b4), no changes at all done on the "server" Fedora machine still running cups-2.3.3op2-11.fc35.

> Can you see the queue from server on mDNS communication? You can check via `avahi-browse -avrt` on the client.

Looks good to me, below is the requested output.
```
$ avahi-browse -avrt
Server version: avahi 0.8; Host name: exherbo.local
E Ifce Prot Name                                          Type                 Domain
+ enp39s0 IPv6 EPSON XP-2100 Series @ remote                _ipp._tcp            local
+ enp39s0 IPv4 EPSON XP-2100 Series @ remote                _ipp._tcp            local
+ enp39s0 IPv6 EPSON XP-2100 Series @ remote                _ipps._tcp           local
+ enp39s0 IPv4 EPSON XP-2100 Series @ remote                _ipps._tcp           local
+ enp39s0 IPv6 EPSON XP-2100 Series @ remote                _printer._tcp        local
+ enp39s0 IPv4 EPSON XP-2100 Series @ remote                _printer._tcp        local
= enp39s0 IPv6 EPSON XP-2100 Series @ remote                _ipp._tcp            local
   hostname = [remote.local]
   address = [2a02:8070:61c0:a00:5e41:3cc5:20af:7627]
   port = [631]
   txt = ["printer-type=0x100E" "printer-state=3" "Color=T" "TLS=1.2" "UUID=4f6ac68c-ce99-3577-411a-9573feaefb9d" "pdl=application/octet-stream,application/pdf,application/postscript,image/jpeg,image/png,image/pwg-raster" "product=(EPSON XP-2100 Series)" "priority=0" "note=remote.localdomain" "adminurl=https://remote.local.:631/printers/XP-2100-Series" "ty=EPSON XP-2100 Series , Epson Inkjet Printer Driver (ESC/P-R) for Linux" "rp=printers/XP-2100-Series" "qtotal=1" "txtvers=1"]
= enp39s0 IPv4 EPSON XP-2100 Series @ remote                _ipp._tcp            local
   hostname = [remote.local]
   address = [192.168.0.220]
   port = [631]
   txt = ["printer-type=0x100E" "printer-state=3" "Color=T" "TLS=1.2" "UUID=4f6ac68c-ce99-3577-411a-9573feaefb9d" "pdl=application/octet-stream,application/pdf,application/postscript,image/jpeg,image/png,image/pwg-raster" "product=(EPSON XP-2100 Series)" "priority=0" "note=remote.localdomain" "adminurl=https://remote.local.:631/printers/XP-2100-Series" "ty=EPSON XP-2100 Series , Epson Inkjet Printer Driver (ESC/P-R) for Linux" "rp=printers/XP-2100-Series" "qtotal=1" "txtvers=1"]
= enp39s0 IPv6 EPSON XP-2100 Series @ remote                _ipps._tcp           local
   hostname = [remote.local]
   address = [2a02:8070:61c0:a00:5e41:3cc5:20af:7627]
   port = [631]
   txt = ["printer-type=0x100E" "printer-state=3" "Color=T" "TLS=1.2" "UUID=4f6ac68c-ce99-3577-411a-9573feaefb9d" "pdl=application/octet-stream,application/pdf,application/postscript,image/jpeg,image/png,image/pwg-raster" "product=(EPSON XP-2100 Series)" "priority=0" "note=remote.localdomain" "adminurl=https://remote.local.:631/printers/XP-2100-Series" "ty=EPSON XP-2100 Series , Epson Inkjet Printer Driver (ESC/P-R) for Linux" "rp=printers/XP-2100-Series" "qtotal=1" "txtvers=1"]
= enp39s0 IPv4 EPSON XP-2100 Series @ remote                _ipps._tcp           local
   hostname = [remote.local]
   address = [192.168.0.220]
   port = [631]
   txt = ["printer-type=0x100E" "printer-state=3" "Color=T" "TLS=1.2" "UUID=4f6ac68c-ce99-3577-411a-9573feaefb9d" "pdl=application/octet-stream,application/pdf,application/postscript,image/jpeg,image/png,image/pwg-raster" "product=(EPSON XP-2100 Series)" "priority=0" "note=remote.localdomain" "adminurl=https://remote.local.:631/printers/XP-2100-Series" "ty=EPSON XP-2100 Series , Epson Inkjet Printer Driver (ESC/P-R) for Linux" "rp=printers/XP-2100-Series" "qtotal=1" "txtvers=1"]
= enp39s0 IPv6 EPSON XP-2100 Series @ remote                _printer._tcp        local
   hostname = [remote.local]
   address = [2a02:8070:61c0:a00:5e41:3cc5:20af:7627]
   port = [0]
   txt = []
= enp39s0 IPv4 EPSON XP-2100 Series @ remote                _printer._tcp        local
   hostname = [remote.local]
   address = [192.168.0.220]
   port = [0]
   txt = []
```

> And to debug this further, it would be great if you could give me debug logs from the client when the forbidden message appears. From the snippet you've shared I'm not sure when the rejection happens, but since you don't see the printer, my guess is it happens during print queue listing, so `lpstat -e` should hit the issue and it shouldn't return anything.

lpstat seems to work fine and displays the remote printer on the client:
```
$ lpstat -e
EPSON_XP_2100_Series_remote
```

> Here are the steps:
> 
> ```
> $ sudo sed -i 's,LogLevel warn,LogLevel debug2,' /etc/cups/cupsd.conf
> $ sudo systemctl restart cups
> $ tail -f /var/log/cups/error_log &> log.txt &
> $ lpstat -e
> ```
> 
> after lpstat finishes, kill the tail process. Please check if log.txt contains the errors you mentioned - if it does, do attach it here as attachment. Otherwise (since you reported the errors occurs regularly) catch the log till the error happens.

Attachment of the error_log: [error_log.txt](https://github.com/OpenPrinting/cups/files/7684343/error_log.txt)

> Thank you in advance!

Huge thanks to you for any help on trying to figure things out!

**Comment by tgurr** [comment]
> The previous comment expects that you are using temporary queues - meaning your client's cupsd picks up queue advertised by mDNS during print dialog and the queue is not permanently installed - which means no cups-browsed is involved.

This is definitely a usefult hint! If I stop cups-browsed.service then the error message stops to appear. I investigated a little bit further and found out that I indeed can print via kwrite or kde native apps, however in Chrome where I initially had the problem I still can't find any printer, same for Firefox, but in Firefox I can then choose "print with system dialogue" where the printer then appears as well. However in the CUPS webinterface the printer isn't visiable, contrary to like it was and still is if I downgrade back to 2.3.3 on the client. I hope this information can give further hints to where the problem might lie, probably mDNS works fine, but cups-browsed doesn't anymore(?).

**Comment by tgurr** [comment]
> I can try to reproduce with them as well, otherwise I would ask to turn to Exherbo Linux for help.

I'm actually maintaining cups on Exherbo Linux hence my strong interest to get this sorted out especially in regards to potential packaging issues. Thanks to your detailed explanations above I was able to further figure things out, for the start we pass `--with-system-groups=lpadmin` during configure which results in 

`/etc/cups/cups-files.conf`

```
# Administrator user group, used to match @SYSTEM in cupsd.conf policy rules...
# This cannot contain the Group value for security reasons...
SystemGroup lpadmin
```

for both 2.3.3op2 and also 2.4.0, the config is identical, however where things work with 2.3.3op2 they don't anymore with 2.4.0, at least not if I do not manually add root (again thanks for the hints!) to the list

`SystemGroup root lpadmin` and/or pass `--with-system-groups="root lpadmin"` to configure which in turn adds the relevant users to the generated configuration file.

I found https://github.com/OpenPrinting/cups/blob/master/config-scripts/cups-defaults.m4#L285 (which also includes "the strange" lpadmin) and there have been some syntax changes between 2.3.3op2 and 2.4.0, could it be the case that the groups listed there have previously just been somehow appended by cupsd to the entry specified in the `cups-files.conf` configuration file which now somehow doesn't work for 2.4.0 anymore? The downstream fix or workaround for this would be that we simply pass `--with-system-groups="lpadmin sys root wheel"` which would result in a changed configuration file

`/etc/cups/cups-files.conf`

```
# Administrator user group, used to match @SYSTEM in cupsd.conf policy rules...
# This cannot contain the Group value for security reasons...
SystemGroup lpadmin sys root wheel
```
however I still don't understand why the same configuration works fine for 2.3.3op2 and starts to fail for 2.4.0, something must have changed.

**Comment by zdohnal** [comment]
@tgurr thanks for the info and checking the code! I'm sorry, you're right, something changed in CUPS and it is quite interesting - I'm able to 'reproduce' the behavior with 'username=''' and 'Unauthorized' and with 'username="root"' and 'Forbidden'.

The former is especially tricky - I get this when I build and install CUPS locally from current master branch, and all add-printer requests from cups-browsed get 'Unauthorized', because incoming packets are missing HTTP Authentication field. The field is populated (to put it simply) in ```cups_local_auth()```, but it seems it isn't. On top of that when I open cups-browsed in gdb to investigate, the field suddenly gets populated properly and adding printer works if I have ```SystemGroup sys root wheel```! More interesting, I cannot reproduce this at all if I put CUPS 2.4.0 rpm package, which I have prepared for Fedora update, which shouldn't have any related patches to the issue...

The latter - 'username=root' and 'Forbidden' appears even with Fedora rpms, if I set 'SystemGroup lpadmin'. configure script logic is the same as before (substituting SystemGroup value with the value from argument), so currently I don't know why it worked as an addition to default system groups before 2.4.0...

Ad the current --with-system-groups behavior - IMHO the current behavior is aligned with other usages of --with arguments - substituting the whole value, not adding the value to other defaults, so it would be great if you updated your configuration to include the necessary groups. Or, if 'lpadmin' already exists on your system, remove the argument as whole - conf script then automatically set system groups with existing groups on your system.

I will try to look deeper into what actually changed which can cause this change.

**Comment by tgurr** [comment]
> so it would be great if you updated your configuration to include the necessary groups. Or, if 'lpadmin' already exists on your system, remove the argument as whole - conf script then automatically set system groups with existing groups on your system.

Thanks for the advice, I've added the suggested changes to our package via:
https://git.exherbo.org/arbor.git/commit/?id=dd3b3aaddc6dae4e541cf2356a452fd689b7fa03
I've also tested building without passing `--with-system-groups` at all which also picked up `lpadmin sys root wheel` and generated a `cups-files.conf` with `SystemGroup lpadmin sys root wheel` but to not rely on any automagic happening I've decided to add them explicitly. Huge thanks for investigating and helping me to figure things out to fix our 2.4.0 package!

**Comment by zdohnal** [comment]
Ok, I've got on the bottom of this - if you set --with-system-groups argument, only the value from this configuration argument is set as SystemGroup, so no 'magic' additions happen further down to the code regarding values in SystemGroup.
What changed was that ```root``` user is no longer automatically allowed, he needs to be in SystemGroup as well - it was introduced by this commit which was part of implementation of Snap support 0cdf79ef5e2dc .
So cups-browsed runs as a root and does request to local cupsd as a root, so once the implicit permission for root was removed, cups-browsed couldn't create a queue locally, because root is not longer implicitly allowed and it wasn't part of lpadmin group.

So the behavior is expected.
